src/Orchard.Environment.Extensions/ExtensionManager.cs

```diff
@@ -1,227 +1,198 @@
+            // Results are cached so that there is no mismatch when loading an assembly twice.
+            // Otherwise the same types would not match.
+
+            try
             {
-                _availableFeatures = AvailableExtensions()
-                    .SelectMany(ext => ext.Features)
-                    .OrderByDependenciesAndPriorities(HasDependency, GetPriority)
-                    .ToList();
-            }
+                return await _extensions.Get(extensionInfo.Id, id =>
+                {
+                    var extension = _extensionLoader.Load(extensionInfo);
+                    
+                    if (extension == null && _logger.IsEnabled(LogLevel.Warning))
+                    {
+                        _logger.LogWarning("No suitable loader found for extension \"{0}\"", extensionInfo.Id);
+                    }
 
-            return _availableFeatures;
+                    return Task.FromResult(extension);
+                });
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(string.Format("Error loading extension '{0}'", extensionInfo.Id), ex);
+                throw new OrchardException(T["Error while loading extension '{0}'.", extensionInfo.Id], ex);
+            }
         }
 
-        internal static int GetPriority(FeatureDescriptor featureDescriptor)
-        {
-            return featureDescriptor.Priority;
-        }
+        public async Task<IEnumerable<ExtensionEntry>> LoadExtensionsAsync(IEnumerable<IExtensionInfo> extensionInfos) {
+            var extensionEntries = new ConcurrentBag<ExtensionEntry>();
 
-        /// <summary>
-        /// Returns true if the item has an explicit or implicit dependency on the subject
-        /// </summary>
-        /// <param name="item"></param>
-        /// <param name="subject"></param>
-        /// <returns></returns>
-        public bool HasDependency(FeatureDescriptor item, FeatureDescriptor subject)
-        {
-            if (DefaultExtensionTypes.IsTheme(item.Extension.ExtensionType))
+            Parallel.ForEach(extensionInfos, async extensionInfo =>
             {
-                if (DefaultExtensionTypes.IsModule(subject.Extension.ExtensionType))
+                try
                 {
-                    // Themes implicitly depend on modules to ensure build and override ordering
-                    return true;
+                    var extension = await LoadExtensionAsync(extensionInfo);
+                    extensionEntries.Add(extension);
                 }
-
-                if (DefaultExtensionTypes.IsTheme(subject.Extension.ExtensionType))
+                catch (Exception e)
                 {
-                    // Theme depends on another if it is its base theme
-                    return item.Extension.BaseTheme == subject.Id;
+                    extensionEntries.Add(new FailedExtensionEntry { Exception = e, ExtensionInfo = extensionInfo });
```

Comment:
The FailedExtensionEntry is just here to pass error messages and make an ugly pattern later on (typeof() == FailedExtensionEntry). So nothing about extensibility. Just remove this class and return a bool to expose success, and this class can log the exception by itself.